pydantic/_internal/_core_utils.py

```diff
@@ -413,42 +415,34 @@ def collect_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.Cor
             for definition in s['definitions']:
                 ref = get_ref(definition)
                 assert ref is not None
-                def_schema = recurse(definition, collect_refs).copy()
-                if 'invalid' in def_schema.get('metadata', {}):
-                    invalid_defs[ref] = def_schema
-                else:
-                    valid_defs[ref] = def_schema
+                all_defs[ref] = recurse(definition, collect_refs)
             return recurse(s['schema'], collect_refs)
-        ref = get_ref(s)
-        if ref is not None:
-            if 'invalid' in s.get('metadata', {}):
-                invalid_defs[ref] = s
-            else:
-                valid_defs[ref] = s
-        return recurse(s, collect_refs)
+        else:
+            ref = get_ref(s)
+            if ref is not None:
+                all_defs[ref] = s
+            return recurse(s, collect_refs)
 
     schema = walk_core_schema(schema, collect_refs)
 
-    all_defs = {**invalid_defs, **valid_defs}
-
     def flatten_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
-        if is_definitions_schema(s):
-            new: dict[str, Any] = dict(s)
+        if s['type'] == 'definitions':
             # iterate ourselves, we don't want to flatten the actual defs!
-            definitions: list[CoreSchema] = new.pop('definitions')
-            schema = cast(CoreSchema, new.pop('schema'))
+            definitions: list[CoreSchema] = s.pop('definitions')  # type: ignore
```

Comment:
It's slightly unfortunate to have the `type: ignore` lines added here, I assume that you removed the `TypeGuard` as a micro-optimization to avoid overhead hitting every branch. Are you willing to consider adding a `def_schema = cast(s, core_schema.DefinitionsSchema)` back here?

Reply:
The `TypeGuard` doesn't help at all here - basically any combination here requires multiple casts or ignores, I'm using ignore comments as this code is quite hot.